Make the "push to build branch" approach a toggle (#119)
inpsyde/reusable-workflows: docs/build-and-push-assets.md

Reviewer: After incorporating the changes, please have your IDE fix the line lengths again. Doing this in the GitHub suggestion UI is a bit hard. 😉

## Where are assets stored

> Two boolean inputs can be used to define branches as assets storage - `USE_BUILT_BRANCH` and `USE_RELEASE_BRANCH`.

Reviewer: ```suggestion
Two boolean inputs can be used to define branches as asset storage: `USE_BUILT_BRANCH` and `USE_RELEASE_BRANCH`.
```

> `USE_BUILT_BRANCH` is used only for push-to-branch events. If true, compiled assets
> will be stored in the branch with name equals current branch plus `-built` suffix. For instance,
> for pushing to the `main` branch, compiled assets will be stored in the `main-built` branch (a branch will be created if
> it does not exist).

Reviewer: ```suggestion
`USE_BUILT_BRANCH` is only used for `branch` events. If set, compiled assets will be stored in the branch whose name corresponds to the current branch plus the `-built` suffix. For example, when pushing to the `main` branch, compiled assets will be stored in the `main-built` branch (a branch will be created if it does not exist).
```

> `USE_RELEASE_BRANCH` is used only for tag events. If true and the pushed tag points
> to the last commit of the default branch of the GitHub repository, compiled assets will be pushed
> to the `release` branch, and the tag will be moved there.
> If the input is undefined or the tag points to one of the previous commits,
> the compiled assets will be pushed to the detached commit, and the tag will be moved there.

Reviewer: ```suggestion
`USE_RELEASE_BRANCH` is only used for tag events. If set, and the tag being pushed points to the latest commit of the default branch of the GitHub repository, compiled assets will be pushed to the `release` branch, and the tag will be moved there.
If the input is undefined or the tag points to one of the previous commits, the compiled assets will be pushed to the detached commit, and the tag will be moved there.
```

The main benefit of using `USE_BUILT_BRANCH` is not to pollute the main development branch
with commits containing compiled assets. With `USE_RELEASE_BRANCH`, you can gain linear tag history

> if you always tag just the last commit from the main development branch.

Reviewer: ```suggestion
by always tagging only the latest commit from the main development branch.
```

  push:
    tags: ['*']
    branches: ['*']

>     # Don't include paths if USE_BUILT_BRANCH or USE_RELEASE_BRANCH are true

Reviewer: ```suggestion
    # Don't include paths if USE_BUILT_BRANCH or USE_RELEASE_BRANCH are set.
```

| `COMPILE_SCRIPT_DEV`  | `'build:dev'`                 | Script added to `npm run` or `yarn` to build development assets                                                                 |
| `MODE`                | `''`                          | Mode for compiling assets (`prod` or `dev`)                                                                                     |
| `ASSETS_TARGET_PATHS` | `'./assets'`                  | Target path(s) for compiled assets                                                                                              |

> | `USE_BUILT_BRANCH`    | `false`                       | On push-to-branch events, to activate a target branch where compiled assets are pushed                                          |
> | `USE_RELEASE_BRANCH`  | `false`                       | On tag events, target `release` branch where compiled assets are pushed and the tag is moved to                                 |

Reviewer: ```suggestion
| `USE_BUILT_BRANCH`    | `false`                       | On `branch` events, push compiled assets to a derived branch with the "-built" suffix                                          |
| `USE_RELEASE_BRANCH`  | `false`                       | On `tag` events, push compiled assets to a specific release branch and move the tag to that branch                               |
```

As for the "noise", it will indeed be there. However, considering that all workflow commit messages
start with the prefix `[BOT]`, it would be quite easy to ignore them without any cognitive effort.

> By defining `USE_BUILT_BRANCH`, you keep commits containing compiled assets separated in the built branch.

Reviewer: ```suggestion
By setting `USE_BUILT_BRANCH`, you keep commits containing compiled assets separated in the built branch.
```

For tags, the pushed tag name is always used.

> For branches, it depends on the `USE_BUILT_BRANCH` input value. If the input is not provided,
> you can use the branch name (i.e., `dev-main` for the `main` branch) as usual. If the input value is true,
> the built branch should be used. I.e., in case of `main` branch, the `dev-main-built` branch should be used as the
> package version.

Reviewer: ```suggestion
For branches, it depends on the value of the `USE_BUILT_BRANCH` input. If it is set, the "built" branch should be used. For example, when working with the `main` branch, require the `dev-main-built` branch. If it is not set, use the branch name as usual (i.e., `dev-main` for the `main` branch).
```
